src/d3a/d3a_core/redis_connections/aggregator_connection.py

```diff
@@ -0,0 +1,84 @@
+import json
+import logging
+from threading import Lock
+import d3a.constants
+
+
+class AggregatorHandler:
+
+    def __init__(self):
+        self.pending_batch_commands = {}
+        self.processing_batch_commands = {}
+        self.responses_batch_commands = {}
+        self.batch_events = {}
+        self.aggregator_device_mapping = {}
+        self.device_aggregator_mapping = {}
+        self.lock = Lock()
+
+    def set_aggregator_device_mapping(self, aggregator_device):
+        self.aggregator_device_mapping = aggregator_device
+        self.device_aggregator_mapping = {
+            dev: aggr
+            for aggr, devices in self.aggregator_device_mapping
+            for dev in devices
+        }
+
+    def is_controlling_device(self, device_uuid):
+        return device_uuid in self.device_aggregator_mapping
+
+    def add_batch_event(self, device_uuid, event):
+        aggregator_uuid = self.device_aggregator_mapping[device_uuid]
+
+        if aggregator_uuid not in self.batch_events:
+            self.batch_events[aggregator_uuid] = []
+
+        self.batch_events[aggregator_uuid].append(event)
+
+    def receive_batch_commands_callback(self, payload):
+        batch_command_message = json.loads(payload["data"])
+        transaction_id = batch_command_message["transaction_id"]
+        with self.lock:
+            self.pending_batch_commands[transaction_id] = batch_command_message
+
+    def approve_batch_commands(self):
```

Comment:
I am having trouble understanding why this additional approval step is needed. In the live events we also "only" have a buffer that is consumed every market slot. Can you help me please? thanks!

Reply:
The extra step is required here exactly for the same reason as the live events. Tick event takes time. Therefore we need to buffer any incoming commands that arrive during the tick to be processed at the next tick. So, at the start of each tick, we move these commands from the 'incoming' buffer to the 'processing' buffer. 
Only difference with the live events, is that this buffer is per-tick, not per-market cycle.

Reply:
OK, understood. What I meant is that in the liveEvents, we have inly one buffer. The `event_buffer` . It gets filled in the frame of one slot and the digested at the beginning of the next slot. Why do we need two buffers (incoming and processing) here?

Reply:
Answered on Slack already, the difference here is that in our case the first buffer does not need to be consumed in one go, since it is consumed as long as the tick event propagates. This adds an extra danger of messages arriving while existing messages are being processed. We do not have this problem in live-events, because we consume all the events in one go, which is protected by a lock.